language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/ClassForType.scala

```diff
@@ -15,21 +15,27 @@ object ClassForType extends StrictLogging {
   def apply(
       typeWithContext: TypeWithContext,
       packagePrefixes: Map[PackageId, String],
+      relevant: Identifier => Boolean,
   ): List[JavaFile] = {
 
-    val classNameString = fullyQualifiedName(typeWithContext.identifier, packagePrefixes)
-    val className = ClassName.bestGuess(classNameString)
-
-    def generate(lfInterfaceType: InterfaceType): List[JavaFile] =
-      generateInterfaceTypes(typeWithContext, packagePrefixes) ++
-        generateConcreteTypes(typeWithContext, className, packagePrefixes, lfInterfaceType)
-
     def recurOnTypeLineages: List[JavaFile] =
       typeWithContext.typesLineages
-        .flatMap(ClassForType(_, packagePrefixes))
+        .flatMap(ClassForType(_, packagePrefixes, relevant))
         .toList
 
-    typeWithContext.`type`.typ.fold(recurOnTypeLineages)(generate)
+    def generateForType(lfInterfaceType: InterfaceType): List[JavaFile] = {
+      val classNameString = fullyQualifiedName(typeWithContext.identifier, packagePrefixes)
+      val className = ClassName.bestGuess(classNameString)
+      generateInterfaceTypes(typeWithContext, packagePrefixes) ++
+        generateSerializableTypes(typeWithContext, className, packagePrefixes, lfInterfaceType)
+    }
+
+    if (relevant(typeWithContext.identifier)) {
+      typeWithContext.`type`.typ.fold(recurOnTypeLineages)(generateForType)
+    } else {
+      recurOnTypeLineages
+    }
```

Comment:
```suggestion
    Option
      .when(relevant(typeWithContext.identifier))(typeWithContext.`type`.typ)
      .flatten
      .fold(recurOnTypeLineages)(generateForType)
```
Avoid invoking the recursion from two different places.